### PR TITLE
Add ability to manage TTL on tables

### PR DESCRIPTION
**Pull request by acron0**
Add the following functions:

[updateTimeToLive](http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/dynamodbv2/AmazonDynamoDBClient.html#updateTimeToLive-com.amazonaws.services.dynamodbv2.model.UpdateTimeToLiveRequest-) as `update-ttl`
[describeTimeToLive](http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/dynamodbv2/AmazonDynamoDBClient.html#describeTimeToLive-com.amazonaws.services.dynamodbv2.model.DescribeTimeToLiveRequest-) as `describe-ttl` 

Notes:
I was required to update the AWS SDK which I can only assume impacted some of the test output. I've tweaked some of the tests without undermining any fundamentals (to the best of my knowledge). I noticed that tests were also failing on a fresh checkout and my alterations meant they were failing for different reasons so I felt obliged to look into it. I believe I've managed to keep all the tests relevant and to the spirit with which they were intended. There are some differences between the local and non-local testing systems which, again, I've tried to cater for.

Tests against local DynamoDB (via `./run-tests local`) failed (it doesn't appear to support the new functions).
```
com.amazonaws.services.dynamodbv2.model.AmazonDynamoDBException: 
   An unknown operation was requested.
```    
As a workaround, I have added the `AWS_DYNAMODB_LOCAL` environment variable into the `run-tests` script for when running against a local instance. It omits the new tests which cause the issue. In my opinion the local version of DynamoDB exhibits such varying behaviour now that deprecation of that testing mechanism should be considered.

Tests against _real_ DynamoDB (via `./run-tests my-env`) failed due to 
```
com.amazonaws.services.dynamodbv2.model.LimitExceededException: 
   Subscriber limit exceeded: Only 1 online index can be created or deleted simultaneously per table
```
From what I can see, this problem is occurring on `master` and was not introduced during my PR. My attempt to solve this is with an improved `index-status-watch` function.

### CODE REVIEW

**Comment by belucid** [comment]
Thanks for the PR @acron0 .

Taking a look at this one now.

File: project.clj [hunk 1]
```diff
@@ -15,7 +15,7 @@
    [com.taoensso/encore "2.88.1"]
    [com.taoensso/nippy  "2.12.2"]
    [joda-time           "2.9.6"]
-   [com.amazonaws/aws-java-sdk-dynamodb "1.10.49"
+   [com.amazonaws/aws-java-sdk-dynamodb "1.11.253"
```
Comment (belucid) [inline project.clj]:
```suggestion
   [com.amazonaws/aws-java-sdk-dynamodb "1.11.462"
```

File: project.clj [hunk 2]
```diff
@@ -25,7 +25,7 @@
    :1.6  {:dependencies [[org.clojure/clojure "1.6.0"]]}
    :1.7  {:dependencies [[org.clojure/clojure "1.7.0"]]}
    :1.8  {:dependencies [[org.clojure/clojure "1.8.0"]]}
-   :1.9  {:dependencies [[org.clojure/clojure "1.9.0-alpha10"]]}
+   :1.9  {:dependencies [[org.clojure/clojure "1.9.0"]]}
```
Comment (belucid) [inline project.clj]:
```suggestion
   :1.9  {:dependencies [[org.clojure/clojure "1.9.0"]]}
   :1.10  {:dependencies [[org.clojure/clojure "1.10.0-RC2"]]}
```

File: src/taoensso/faraday.clj [hunk 3]
```diff
@@ -260,6 +266,11 @@
     (BigDecimal. s)
     (bigint (BigInteger. s))))
 
+(defn nil->zero
```
Comment (belucid) [inline src/taoensso/faraday.clj]: Seems like overkill. Won't `or` do the same thing?

e.g.

```
:item-count (or (.getItemCount d) 0)
```

File: src/taoensso/faraday.clj [hunk 4]
```diff
@@ -421,8 +432,8 @@
   (as-map [d]
     {:name                (keyword (.getTableName d))
      :creation-date       (.getCreationDateTime d)
-     :item-count          (.getItemCount d)
-     :size                (.getTableSizeBytes d)
+     :item-count          (nil->zero (.getItemCount d))
```
Comment (belucid) [inline src/taoensso/faraday.clj]:
```suggestion
     :item-count          (or (.getItemCount d) 0)
```

File: src/taoensso/faraday.clj [hunk 5]
```diff
@@ -421,8 +432,8 @@
...
+     :size                (nil->zero (.getTableSizeBytes d))
```
Comment (belucid) [inline src/taoensso/faraday.clj]:
```suggestion
     :size                (or (.getTableSizeBytes d) 0)
```

File: src/taoensso/faraday.clj [hunk 6]
```diff
@@ -453,16 +464,17 @@
   LocalSecondaryIndexDescription
   (as-map [d]
     {:name       (keyword (.getIndexName d))
-     :size       (.getIndexSizeBytes d)
-     :item-count (.getItemCount d)
+     :size       (nil->zero (.getIndexSizeBytes d))
```
Comment (belucid) [inline src/taoensso/faraday.clj]:
```suggestion
     :size       (or (.getIndexSizeBytes d) 0)
```

File: src/taoensso/faraday.clj [hunk 7]
```diff
@@ -453,16 +464,17 @@
...
+     :item-count (nil->zero (.getItemCount d))
```
Comment (belucid) [inline src/taoensso/faraday.clj]:
```suggestion
     :item-count (or (.getItemCount d) 0)
```

File: src/taoensso/faraday.clj [hunk 8]
```diff
@@ -453,16 +464,17 @@
...
      :key-schema (as-map (.getKeySchema d))
      :projection (as-map (.getProjection d))})
 
   GlobalSecondaryIndexDescription
   (as-map [d]
     {:name       (keyword (.getIndexName d))
-     :size       (.getIndexSizeBytes d)
-     :item-count (.getItemCount d)
+     :size       (nil->zero (.getIndexSizeBytes d))
```
Comment (belucid) [inline src/taoensso/faraday.clj]:
```suggestion
     :size       (or (.getIndexSizeBytes d) 0)
```

File: src/taoensso/faraday.clj [hunk 9]
```diff
@@ -453,16 +464,17 @@
...
+     :item-count (nil->zero (.getItemCount d))
```
Comment (belucid) [inline src/taoensso/faraday.clj]:
```suggestion
     :item-count (or (.getItemCount d) 0)
```

File: src/taoensso/faraday.clj [hunk 10]
```diff
@@ -473,7 +485,7 @@
      :write               (.getWriteCapacityUnits d)
      :last-decrease       (.getLastDecreaseDateTime d)
      :last-increase       (.getLastIncreaseDateTime d)
-     :num-decreases-today (.getNumberOfDecreasesToday d)})
+     :num-decreases-today (nil->zero (.getNumberOfDecreasesToday d))})
```
Comment (belucid) [inline src/taoensso/faraday.clj]:
```suggestion
     :num-decreases-today (or (.getNumberOfDecreasesToday d) 0)})
```

**Comment by belucid** [comment]
Indeed local DynamoDB doesn't support TTL.

**Comment by belucid** [comment]
I didn't experience "tests were also failing on a fresh checkout".

On a fresh checkout and a run of `./run-tests local` I get all tests passing.

**Comment by belucid** [comment]
@acron0 do drop me a note here to let me know you're still interested in getting this PR in. Given my initial review, it's probably close, we'll just need a few rounds of back and forth.

If you're not still interested, that's OK too. With your permission I can pull your work in and adjust it myself (with full credit to you of course). Just let me know either way.

Thanks again.

**Comment by acron0** [comment]
@belucid Heya, thanks for picking this up. Interestingly I no longer have access to the repo where the branch exists, so I cannot commit your in-line changes :( But I would do, if I could. I think the changes could still be useful so I think they should still go in. I'm not in a fantastic position to pull out the changes into a new repo and re-create the PR (work, uh) so I would be grateful if you'd take on the integration work. Thanks.

**Comment by belucid** [comment]
Yup, totally happy to do that @acron0 

I'll of course still attribute the contributions to you.

Thanks for putting this together, and sorry it took so long to act on.

**Comment by joelittlejohn** [comment]
I've taken snippets from this PR and reworked a little onto this branch: https://github.com/Taoensso/faraday/tree/ttl

Still WIP but I thought I'd mention this so that we don't duplicate the work.

**Comment by joelittlejohn** [comment]
Since this is somewhat out-of-date I'll open a new PR when the time comes.

**Pull request closed without merging**